clients/py/sapphirepy/sapphire.py

```diff
@@ -220,58 +251,175 @@ def sapphire_middleware(
+                            )["hash"].hex(),
+                            "gas_price": self._w3.to_wei(
+                                params[0].get("gasPrice", DEFAULT_GAS_PRICE), "wei"
+                            ),
+                        }
+                        c = _encrypt_tx_params(
+                            pk, params, self.sapphire_account, signed_call_data
+                        )
+                    else:
+                        c = _encrypt_tx_params(pk, params, self.sapphire_account)
 
                     # We may encounter three errors here:
                     #  'core: invalid call format: epoch too far in the past'
                     #  'core: invalid call format: Tag verification failed'
                     #  'core: invalid call format: epoch in the future'
                     # We can only do something meaningful with the first!
                     result = cast(RPCResponse, make_request(method, params))
-                    if result.get('error', None) is not None:
-                        error = result['error']
-                        if not isinstance(error, str) and error['code'] == -32000:
-                            if error['message'] == 'core: invalid call format: epoch too far in the past':
+                    if result.get("error", None) is not None:
+                        error = result["error"]
+                        if not isinstance(error, str) and error["code"] == -32000:
+                            if (
+                                error["message"]
+                                == "core: invalid call format: epoch too far in the past"
+                            ):
                                 # force the re-fetch, and encrypt with new key
                                 do_fetch = True
                                 pk = None
                                 continue
 
                 # Only eth_call is decrypted
-                if method == 'eth_call' and result.get('result', '0x') != '0x':
-                    decrypted = c.decrypt(unhexlify(result['result'][2:]))
-                    result['result'] = HexStr('0x' + hexlify(decrypted).decode('ascii'))
+                if method == "eth_call" and result.get("result", "0x") != "0x":
+                    decrypted = c.decrypt(unhexlify(result["result"][2:]))
+                    result["result"] = HexStr("0x" + hexlify(decrypted).decode("ascii"))
 
                 return result
             return make_request(method, params)
 
         return middleware
 
-    return sapphire_middleware
+    async def async_wrap_make_request(
+        self, make_request: "AsyncMakeRequestFn"
+    ) -> "AsyncMakeRequestFn":
+        """
+        Async version of wrap_make_request that handles the same encryption logic
+        for eth_estimateGas, eth_sendTransaction, and eth_call
+        """
+        manager = CalldataPublicKeyManager()
+
+        async def middleware(method: "RPCEndpoint", params: Any) -> "RPCResponse":
```

Comment:
This function is essentialyl the same as the non-async `middleware()` above. Let's deduplicate this code. Perhaps use the async version by default and put a non-async caller in front of it.

Reply:
I added a possible solution, "wrap_make_request" now calls the async middleware in a separate thread and waits for execution to end, mimicking synchronous call. "async_wrap_make_request" coroutine fetches block_number/chain_id/transaction_count/block_hash either as await or normal function depending on web3 type (Web3/AsyncWeb3).